### ISSUE TITLE
ICE: Implementing trait with associated constant and `&Self` bound

### ISSUE THREAD

**Issue by madsmtm** [rust-lang/rust#88599]
### Code

```Rust
pub trait First {
    const CONST: bool;
}
pub trait Second {}

impl<'a> First for dyn Second where &'a Self: First {
    const CONST: bool = <&Self>::CONST;
}
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7aadd0b8541b302a6e2992f485649630).

### Meta
Tested on rust `1.54.0` and nightly `2021-08-31`.

`rustc --version --verbose`:
```
rustc 1.54.0 (a178d0322 2021-07-26)
binary: rustc
commit-hash: a178d0322ce20e33eac124758e837cbd80a6f633
commit-date: 2021-07-26
host: x86_64-apple-darwin
release: 1.54.0
LLVM version: 12.0.1
```

### Error output

```
error: internal compiler error: Encountered error `Unimplemented` selecting `Binder(<&dyn Second as First>, [])` during codegen
  |
  = note: delayed at compiler/rustc_trait_selection/src/traits/codegen.rs:68:32

error: internal compiler error: ty::ConstKind::Error constructed but no error reported.
  |
  = note: delayed at compiler/rustc_middle/src/ty/consts.rs:183:43

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler/rustc_errors/src/lib.rs:1166:13
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0: rust_begin_unwind
             at library/std/src/panicking.rs:515:5
   1: std::panicking::begin_panic_fmt
             at library/std/src/panicking.rs:457:5
   2: rustc_errors::HandlerInner::flush_delayed
   3: <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop
   4: core::ptr::drop_in_place<rustc_session::parse::ParseSess>
   5: <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop
   6: core::ptr::drop_in_place<rustc_interface::interface::Compiler>
   7: rustc_span::with_source_map
   8: rustc_interface::interface::create_compiler_and_run
   9: scoped_tls::ScopedKey<T>::set
```

</p>
</details>

**Comment by SNCPlay42** [comment]
This version without `dyn Trait` also ICEs ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=408d8e98a385a6763fe6dd53eff60393)):

```rust
pub trait Foo {
    const CONST: bool;
}

impl <'a> Foo for () where &'a Self: Foo {
    const CONST: bool = <&Self>::CONST;
}
```

**Comment by SNCPlay42** [comment]
Both versions compile on Rust 1.43.0: [godbolt](https://rust.godbolt.org/z/fq45GYhc5)

@rustbot label regression-from-stable-to-stable

**Comment by apiraino** [comment]
probably too old to bisect but it seems introduced in rollup #70371 on nightly-2020-03-26

**Comment by apiraino** [comment]
Assigning priority as discussed in the [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2388599.20ICE.3A.20Implementing.20.60dyn.20Trait.60.20with.20Self.20bound.20and.20as.E2.80.A6) of the Prioritization Working Group.

@rustbot label -I-prioritize +P-medium
